File: ballast-core/npm/src/domain/models/vessel.ts
import { CubicCoordinates } from "./cubic-coordinates";
import { Player } from "./player";
import { VesselRole } from "./vessel-role";

export class Vessel {

    public readonly id: string;
    public name: string;
    public cubicCoordinates: CubicCoordinates;
    public captain: Player | null;
    public radioman: Player | null;

    private constructor(id: string, name: string, cubicCoordinates: CubicCoordinates, captain: Player | null = null, radioman: Player | null = null) {
        this.id = id;
        this.name = name;
        this.cubicCoordinates = cubicCoordinates;
        this.captain = captain || null;
        this.radioman = radioman || null;
    }

    public updateCoordinates(cubicCoordinates: CubicCoordinates) {
        this.cubicCoordinates = cubicCoordinates;
        return this.cubicCoordinates;
    }
    
    public setVesselRole(vesselRole: VesselRole, player: Player | null) {
        if (vesselRole.equals(VesselRole.Captain))
        {
            this.captain = player;
            return;
        }
        if (vesselRole.equals(VesselRole.Radioman))
        {
            this.radioman = player;
            return;
        }
        throw new Error(`Vessel role matching value '${vesselRole.value}' does not exist on the current vessel`);
    }

    public removePlayer(player: Player) {
        if (!player)
            throw new Error("Player cannot be null");
        if (this.captain && this.captain.id == player.id)
            this.captain = null;
        if (this.radioman && this.radioman.id == player.id)
            this.radioman = null;
    }

}

File: ballast-core/dotnet/src/domain/models/Vessel.cs
using System;
using System.Collections.Generic;

namespace Ballast.Core.Domain.Models
{
    public class Vessel
    {

        public Guid Id { get; private set; }
        public string Name { get; private set; }
        public CubicCoordinates CubicCoordinates { get; private set; }
        public Player Captain { get; private set; }
        public Player Radioman { get; private set; }

        public Vessel(Guid id, string name, CubicCoordinates cubicCoordinates, Player captain, Player radioman)
        {
            Id = id;
            Name = name;
            CubicCoordinates = cubicCoordinates;
            Captain = captain;
            Radioman = radioman;
        }

        public CubicCoordinates UpdateCoordinates(CubicCoordinates cubicCoordinates)
        {
            CubicCoordinates = cubicCoordinates;
            return CubicCoordinates;
        }

        public void UpdateName(string name)
        {
            Name = name;
        }

        public void SetVesselRole(VesselRole vesselRole, Player player)
        {
            if (vesselRole.Equals(VesselRole.Captain))
            {
                Captain = player;
                return;
            }
            if (vesselRole.Equals(VesselRole.Radioman))
            {
                Radioman = player;
                return;
            }
            throw new KeyNotFoundException($"Vessel role matching value '{vesselRole.Value}' does not exist on the current vessel");
        }

        public void RemovePlayer(Player player) 
        {
            if (player == null || player.Id == default(Guid))
                throw new ArgumentNullException(nameof(player.Id));
            if (Captain != null && Captain.Id == player.Id)
                Captain = null;
            if (Radioman != null && Radioman.Id == player.Id)
                Radioman = null;
        }

    }
}